src/com/ms/silverking/cloud/dht/management/aws/Util.java

```diff
@@ -21,377 +21,373 @@
+  }
+
+  static List<String> getInstanceIds(List<Instance> instances) {
+    List<String> ids = new ArrayList<>();
+
+    for (Instance instance : instances)
+      ids.add(instance.getInstanceId());
+
+    return ids;
+  }
+
+  static List<String> getIps(List<Instance> instances) {
+    List<String> ips = new ArrayList<>();
+
+    for (Instance instance : instances)
+      ips.add(instance.getPrivateIpAddress());
+
+    return ips;
+  }
+
+  static void writeToFile(String filename, String content) {
+    File file = new File(filename);
+
+    try {
+      // mkdirs and createNewFile only create iff those paths/file don't already exist. so no need to check for
+      // existence beforehand.
+      file.getParentFile().mkdirs();
+      file.createNewFile();
+      FileWriter writer = new FileWriter(file);
+      writer.write(content);
+      writer.close();
+    } catch (IOException e) {
+      e.printStackTrace();
     }
-    
-    private static void sleep(long seconds) {
-        try {
-            Thread.sleep(seconds*1_000);
-        } 
-        catch (InterruptedException ex) {
-            throwRuntimeException(ex);
-        }
+
+    // needs to be done in this order, b/c "everyone else" wipes out all permissions (including "owner")...
+    // "everyone else"
+    file.setExecutable(false, false);
+    file.setReadable(false, false);
+    file.setWritable(false, false);
+
+    // "owner"
+    file.setExecutable(false);
+    file.setReadable(true);
+    file.setWritable(true);
+
+    //        System.out.println("Is Execute allow : " + file.canExecute());
+    //        System.out.println("Is Write allow : " +   file.canWrite());
+    //        System.out.println("Is Read allow : " +    file.canRead());
+  }
+
+  public static void writeToFile(String filename, List<String> content) {
+    writeToFile(filename, String.join(NEW_LINE, content) + NEW_LINE);
+  }
+
+  public static List<String> readFile(String filename) {
+    try {
+      return Files.readAllLines(Paths.get(filename));
+    } catch (IOException ex) {
+      throw new RuntimeException(ex);
     }
-    
-    private static long printMinutesElapsed(long sleepSeconds, int retries, long lastMinutePrinted) {
-        long minute = sleepSeconds*retries / 60;
-        if (minute != 0 && minute != lastMinutePrinted) {
-            System.out.println("      *" + minute + " mins elapsed*");
-            lastMinutePrinted = minute;
+  }
+
+  static void waitForInstancesToBeRunning(AmazonEC2 ec2, List<Instance> instances) {
+    print("  Waiting for Instances to be running");
+
+    DescribeInstancesRequest diRequest = new DescribeInstancesRequest();
+    diRequest.withInstanceIds(getInstanceIds(instances));
+
+    long lastMinutePrinted = 0;
+    long sleepSeconds = 5;
+    long totalRunTimeSeconds = 2 * 60;
+    int retriesCount = 0;
+    List<String> ips = getIps(instances);
+    while (!ips.isEmpty()) {
+      DescribeInstancesResult response = ec2.describeInstances(diRequest);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code appears to be waiting for a resource before it runs. You could use the waiters feature to help improve efficiency. Consider using InstanceExists, InstanceRunning, InstanceStopped or InstanceTerminated. For more information, see [https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/](https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/)